lib/active_merchant/billing/gateways/decidir_plus.rb

```diff
@@ -94,26 +94,53 @@ def add_reference(authorization)
-        if payment.is_a?(String)
+        return unless payment
+
+        case payment
+        when NetworkTokenizationCreditCard
+          add_common_payment_data(post, payment, options)
+          add_network_token(post, payment, options)
+        when String
           token, bin = payment.split('|')
-          post[:token] = token
-          post[:bin] = bin
+          post[:token] = token || ''
+          post[:bin] = bin || ''
         else
+          add_common_payment_data(post, payment, options)
           post[:card_number] = payment.number
           post[:card_expiration_month] = format(payment.month, :two_digits)
           post[:card_expiration_year] = format(payment.year, :two_digits)
           post[:security_code] = payment.verification_value.to_s
-          post[:card_holder_name] = payment.name.empty? ? options[:name_override] : payment.name
-          post[:card_holder_identification] = {}
-          post[:card_holder_identification][:type] = options[:card_holder_identification_type] if options[:card_holder_identification_type]
-          post[:card_holder_identification][:number] = options[:card_holder_identification_number] if options[:card_holder_identification_number]
-
-          # additional data used for Visa transactions
-          post[:card_holder_door_number] = options[:card_holder_door_number].to_i if options[:card_holder_door_number]
-          post[:card_holder_birthday] = options[:card_holder_birthday] if options[:card_holder_birthday]
         end
       end
 
+      def add_common_payment_data(post, payment, options)
+        card_data = payment.is_a?(NetworkTokenizationCreditCard) ? (post[:card_data] = {}) : post
+        card_data[:card_holder_name] = payment.name.empty? ? options[:name_override] : payment.name
+        card_data[:card_holder_identification] = {}
+        card_data[:card_holder_identification][:type] = options[:card_holder_identification_type] if options[:card_holder_identification_type]
+        card_data[:card_holder_identification][:number] = options[:card_holder_identification_number] if options[:card_holder_identification_number]
+
+        # additional data used for Visa transactions
+        card_data[:card_holder_door_number] = options[:card_holder_door_number].to_i if options[:card_holder_door_number]
+        card_data[:card_holder_birthday] = options[:card_holder_birthday] if options[:card_holder_birthday]
+      end
+
+      def add_network_token(post, payment, options)
+        post[:is_tokenized_payment] = 'true'
+        post[:fraud_detection] ||= {}
+        post[:fraud_detection][:sent_to_cs] = 'false'
+        post[:card_data] ||= {}
+        post[:card_data][:last_four_digits] = options[:last_4]
+        post[:bin] = payment.number[0..5]
+
+        post[:token_card_data] = {
+          token: payment.number,
+          eci: payment.eci,
+          cryptogram: payment.payment_cryptogram,
+          device_type: options[:device_type]&.to_s
```

Comment:
I would say not add this unless you are going to create a GSF for it in Core

Reply:
I see that it says it's required in the Docs. Once remote tests start working could you try running the tests without device_type. I know it's says required but we us NT on the regular Decidir without issues.

Reply:
What are the possible options for device_type? We might be able to infer it

Reply:
1(Tablet), 2(Phone)& 3(PC)

Reply:
If AP/GP -> phone otherwise PC?

Reply:
I'm not sure but maybe. It also says it's only for mastercard